lib/reporting/report_grouper.rb

```diff
@@ -0,0 +1,158 @@
+      result
+    end
+
+    def build_tree(datas, remaining_rules)
+      return datas if remaining_rules.empty?
+
+      rules = remaining_rules.clone
+      group_and_sort(rules.delete_at(0), rules, datas)
+    end
+
+    def group_and_sort(rule, remaining_rules, datas)
+      result = []
+      groups = group_data_with_rule(datas, rule)
+      sorted_groups = sort_groups_with_rule(groups, rule)
+
+      sorted_groups.each do |group_value, group_datas|
+        result << {
+          is_group: true,
+          header: build_header(rule, group_value, group_datas),
+          header_class: rule[:header_class],
+          summary_row: build_summary_row(rule, group_value, group_datas),
+          summary_row_class: rule[:summary_row_class],
+          data: build_tree(group_datas, remaining_rules)
+        }
+      end
+      result
+    end
+
+    def group_data_with_rule(datas, rule)
+      datas.group_by { |data|
+        if rule[:group_by].is_a?(Symbol)
+          data.full_row[rule[:group_by]]
+        else
+          rule[:group_by].call(data.item, data.full_row)
+        end
+      }
+    end
+
+    def sort_groups_with_rule(groups, rule)
+      groups.sort_by do |group_key, _items|
+        # By default sort with the group_key if no sort_by rule is present
+        if rule[:sort_by].present?
+          rule[:sort_by].call(group_key)
+        else
+          # downcase for better comparaison
+          group_key.is_a?(String) ? group_key.downcase : group_key
+        end
+      end.to_h
+    end
+
+    def build_header(rule, group_value, group_datas)
+      return if rule[:header].blank?
+
+      rule[:header].call(group_value, group_datas.map(&:item), group_datas.map(&:full_row))
+    end
+
+    def build_summary_row(rule, group_value, datas)
+      return if rule[:summary_row].blank?
+
+      proc_args = [group_value, datas.map(&:item), datas.map(&:full_row)]
+      row = rule[:summary_row].call(*proc_args)
+      row = slice_row_fields(OpenStruct.new(row.reverse_merge!(blank_row)))
+      add_summary_row_label(row, rule, proc_args)
+    end
+
+    def add_summary_row_label(row, rule, proc_args)
+      previous_key = nil
+      label = rule[:summary_row_label]
+      label = label.call(*proc_args) if label.respond_to?(:call)
+      # Adds Total before first non empty column
+      row.each_pair do |key, value|
+        row[previous_key] = label and break if value.present?
+
+        previous_key = key
+      end
+      row
+    end
+
+    def blank_row
+      report.columns.transform_values { |_v| "" }
+    end
+
+    def slice_row_fields(row)
+      result = row.clone
+      report.fields_to_hide.each { |field| result.delete_field!(field) }
+      result
+    end
+
+    # Compute the query result item into a result row
+    # We use OpenStruct to it's easier to access the properties
+    # i.e. row.my_field, rows.sum(&:quantity)
+    def build_row(item)
+      OpenStruct.new(report.columns.transform_values do |column_constructor|
```

Comment:
I'm not sure if it matters on this scale but there are some doubts about the performance of OpenStruct:
https://docs.rubocop.org/rubocop-performance/cops_performance.html#performanceopenstruct

Nowadays, you can also do `rows.pluck(:quantity).sum` to achieve the same.

Reply:
Oh yes good point ! good idea re using `pluck`. I was a bit hesitating introducing this `OpenStruct`. Let wait for other input maybe? cause it will take a bit of time to refactor this :)